### ISSUE TITLE
Pivot Tables Implementation

### ISSUE THREAD

**Issue by paulrosenzweig** [metabase/metabase#13662]
This tracks work towards adding a new pivot table visualization in v0.38.0.

### Do first

- [x] Add collapsing/expanding rows - https://github.com/metabase/metabase/issues/14171 - PR: https://github.com/metabase/metabase/pull/14320
- [x] Add subtotals for grouping rows beyond the first - https://github.com/metabase/metabase/issues/14245

### Then
- [x] Add settings to each field in order to access existing field formatting options, as well as sorting - https://github.com/metabase/metabase/issues/14172
- [x] Implement cell actions/drill menus - https://github.com/metabase/metabase/issues/14173
- [x] Implement header actions/drill menus - https://github.com/metabase/metabase/issues/14173
- [x] Fix high priority bugs:
   - [x] Severe performance issues with queries that have 3+ metrics and 3+ dimensions: https://github.com/metabase/metabase/issues/14336 — PR up: https://github.com/metabase/metabase/pull/14448
   - [x] Horizontal scrollbar causing rows to be misaligned: https://github.com/metabase/metabase/issues/14352
   - [x] Numbers should be right-aligned: https://github.com/metabase/metabase/issues/14351 — fixed in https://github.com/metabase/metabase/pull/14448/commits/fa75632414d526ff17645512621986c9913576d5
   - [x] Can't see grand totals in dashboard cards if it scrolls: https://github.com/metabase/metabase/issues/14246
   - [x] Alignment of cell values looks broken: https://github.com/metabase/metabase/issues/14574
   - [x] Custom columns breaks pivot tables: https://github.com/metabase/metabase/issues/14604
   - [x] Filter by attribute drill through displays as a table not a pivot: https://github.com/metabase/metabase/issues/14632
   - [x] Pivot tables broken on dashboard after resize: https://github.com/metabase/metabase/issues/14591

### Finally / follow-ups
- [x] Add dynamic column sizing to help column titles fit a bit better - https://github.com/metabase/metabase/issues/14350 (punt to 0.38.1)
- [x] Dark mode CSS - https://github.com/metabase/metabase/issues/14326

### Future
- Pushdown to Redshift
- Pushdown to BigQuery
- Pushdown to PostgreSQL

## ETA
1/15/2021

---

## Links
- [product doc](https://docs.google.com/document/d/1ScNpNjjYlsi6syyHFKctTlZ149g5ulTIHFnAmmq6h58/edit)
- [eng doc](https://github.com/metabase/metabase/issues/13573)
- feature branch: [`0.38-pivot-tables`](https://github.com/metabase/metabase/tree/0.38-pivot-tables)
- Issue links: #8572

<details>
  <summary>Frontend work</summary>
- [x] Multi-level pivots (new viz type, associated settings) - #13725
- [x] Draggable settings - #13760
- [x] Virtualized table with scroll-synced grids, basic styling - #13879
- [x] Use pivot table endpoints to display subtotals - #13947
- [x] Dashboard click behavior - #14062
  This won't include cell formatting like our table visualization. Subtotals won't be clickable.
- [ ] Contracting/expanding rows - #14320
- [ ] Multiple subtotal rows
- [ ] Collapse/expand all by column
- [ ] Body cell click actions
- [ ] Design polish - #14064
- [x] Tests - #14051
</details>

<details>
<summary>Backend work</summary>
- [x] API `/api/advanced_computation/pivot/dataset` - metabase/metabase#13946
- [x] API `/api/advanced_computation/pivot/card/:card-id/query` metabase/metabase#13970
- [ ] public / embed endpoint
- [x] only query for displayed subtotals
- [x] bitmask for the grouping column
</details>

---

## How to test this feature out
1. Ask a new simple question about Orders from the Sample Dataset.
2. Summarize count and sum of Subtotal, and group by Order.Created_At:Year, User.State, and User.Source.
3. Change visualization to Pivot Table.
4. You should see a sidebar with sections to assign fields to generating table rows, columns, and values. User.State and Order.Created_At should be in the rows; User.Source should be in the columns; and the Count and Sum should be in the values section.
5. In the table itself, you should see columns for User.State and Created_At, with the years repeated for each state. In the settings sidebar, drag the Created_At item up above User.State to switch their grouping order. The table should reload, now showing you the states repeated for each year.
6. As you scroll down the table, you should see a Totals row for each year, showing you the total Count and total Sum for each User.Source.
7. Next, in the settings side panel, try dragging User.State into the “Columns” bucket, above User.Source. The table should reload, resulting in a very wide table, with rows for each year, and columns grouped by each state, with all five sources listed per state.
8. Scroll all the way over to the far right, and you should see the Row totals area, with one column for Count and one for the Sum of Subtotal. The Created At column with the years should be “sticky” and stay in place as you scrolled to the right.
9. Now in the settings, move User.Source and User.State up into the “rows” bucket so that you have Created At, then User.Source, then User.State. After the table reloads, scroll down and you should see a subtotal row for each source, like Affiliate, Facebook, etc., then a subtotal for each year after the fifth source subtotal.
10. Move User.State back to the “columns” bucket, then save the question and add it to a dashboard. Resize the dashboard card to whatever size you’d like, then save the dashboard. Try scrolling the pivot table horizontally and vertically, and you should be able to see the grand total row at the bottom, and the row totals section at the far right.

**Comment by rlotun** [comment]
**Status**: as of the product pod planning meeting today it seems that this was still roughly on track, though it will be tight.  In terms of the backend, the goal is to get the general API implementation done, and prioritize the optimized support for two DBs first: Redshift and Bigquery, and figure out what else can be done in a prioritized list that can span 38 and 39.

**Comment by mazameli** [comment]
Here are things @kdoh and I noticed today in walking through this feature:

**Bugs:**
- In the Summarize sidebar, when hovering over a column in the group-by list, the plus button to add a grouping is missing for most columns, but not all.
- In the visualization settings sidebar, moving a pivoted field from the "columns" bucket to the "rows" bucket broke the viz after query re-run. Also breaks after browser refresh.
- Moving the field that's pivoted in the settings by default (i.e., the field in the "columns" bucket) to the "rows" bucket results in cells missing in the table.
- In Chrome/MS Edge, the slide-in/out transition for the settings sidebar takes ages. Seems to be especially the case with around 3 metrics and 3 dimensions.
- We were seeing double scrollbars when trying to scroll horizontally, and it seemed like we had to scroll down vertically before we could scroll horizontally.

**UX**
- When scrolling down, there's a bit of lag with the frozen row headings 
- When I scroll horizontally, the row headings should be frozen.
- Default column width feels too narrow, and I can't resize the column widths.
- Numeric values in cells should all be right-aligned.

**Comment by mazameli** [comment]
Some more things @kdoh and I saw:

1. The vertical stroke separating the grouping row headings from the value cells shouldn't continue past the Grand Totals area:

<img width="195" alt="image" src="https://user-images.githubusercontent.com/2223916/102289385-6062e100-3ef3-11eb-82b2-133cfd5de8e1.png">

2. Grouping fields should have the ability to click on them in the settings to expand them and change their sort order, and turn their subtotals off/on.

3. Relatedly, we should add a hover state to the field boxes in the sidebar.

4. I'm looking at this on master, so maybe there's another branch that address this, but we have a bit of work to do to make the dragging interaction with fields in the sidebar clearer:
- There should be a "drop zone" where I can drop fields I'm dragging.
- The field being dragged should stay in place, and the “drag fields here” drop zone area should remain and be highlighted until you mouse up.

This is what I'm seeing on master for reference:

![dragging-ix](https://user-images.githubusercontent.com/2223916/102289670-129aa880-3ef4-11eb-9619-6fda42ee7010.gif)

5. There are some alignment problems with column headings. They should be aligned at their baseline, and it looks to me like the metric headings in particular are too high up and too close to the border above them:

<img width="552" alt="image" src="https://user-images.githubusercontent.com/2223916/102289825-6f965e80-3ef4-11eb-8ab4-9a4c969428c4.png">

6. We've already mentioned the scroll lag before, but here's a visual. It seems to happen once I have around 2 metrics and 3 breakouts, with one of those breakouts pivoted:

![scroll-lag](https://user-images.githubusercontent.com/2223916/102290016-fb0fef80-3ef4-11eb-910e-02cf02e8973d.gif)

**Comment by mazameli** [comment]
FYI to all, I added steps for how to manually test this feature locally in the [description](https://github.com/metabase/metabase/issues/13662#issue-734904322) up top.

**Comment by mazameli** [comment]
@paulrosenzweig and I talked prioritization earlier today. The main gaps and priorities that I see are listed below, in a rough sequencing. 

Would appreciate getting @salsakran's two cents — anything look out of order, or anything seem like it could be punted? I've tried to cull it down to the items that are required to keep the feature from having weird gaps.

1. Add collapsing/expanding rows.
2. Add subtotals for grouping rows beyond the first (right now there are only subtotals for the topmost grouping row)
3. Add settings to each field in order to access existing field formatting options, as well as sorting.
4. Implement cell actions/drill menus.
5. Implement header actions/drill menus.
6. Fix high priority bugs
7. Stretch: add dynamic column sizing to help column titles fit a bit better.

@paulrosenzweig and I had initially talked about the possibility of punting #2, but agreed that it would leave weird gaps and limit the usefulness of collapsable rows. Paul estimates implementing it will take "days, not weeks. "I think it’s a < 1hr backend change. The frontend is tricky mostly because I have to change the representation of the left header. not a huge challenge at all, but might have some hiccups."

Here's a visual for item #2 — right now if you have more than 2 grouping rows, you'll only see subtotals for the topmost parent grouping. In this screenshot, you'd expect to see subtotals after each state, but you don't:

<img width="531" alt="image" src="https://user-images.githubusercontent.com/2223916/103594124-9cfb7a00-4eac-11eb-88a9-f46173a0cdc0.png">
